### PR TITLE
Improve configuration parameters for Kafka #1359

### PR DESCRIPTION
**Pull request by chandresh-pancholi**
Signed-off-by: chandresh-pancholi <chandreshpancholi007@gmail.com>

## Which problem is this PR solving?
- https://github.com/jaegertracing/jaeger/issues/1359

## Short description of the changes
- I have added two Kafka config. CompressionLevel, RequiredAck

### CODE REVIEW

File: plugin/storage/kafka/options.go [hunk 1]
```diff
@@ -36,10 +36,14 @@ const (
 	suffixBrokers  = ".brokers"
 	suffixTopic    = ".topic"
 	suffixEncoding = ".encoding"
+	suffixRequiredAcks = ".requiredacks"
```
Comment (objectiser) [inline plugin/storage/kafka/options.go]: nit: Personally would prefer `.required.acks` - similar to the `request.required.acks` mentioned in the godocs.

File: plugin/storage/kafka/options.go [hunk 2]
```diff
@@ -69,12 +73,24 @@ func (opt *Options) AddFlags(flagSet *flag.FlagSet) {
 		defaultEncoding,
 		fmt.Sprintf(`(experimental) Encoding of spans ("%s" or "%s") sent to kafka.`, EncodingJSON, EncodingProto),
 	)
+	flagSet.Int(
+		configPrefix+suffixRequiredAcks,
+		defaultRequiredAcks,
+		"(experimental) Required kafka broker acknowledgement",
```
Comment (objectiser) [inline plugin/storage/kafka/options.go]: Would be good to summarise the valid values and what is the default.

File: plugin/storage/kafka/options.go [hunk 3]
```diff
@@ -69,12 +75,29 @@ func (opt *Options) AddFlags(flagSet *flag.FlagSet) {
...
+	)
 }
 
 // InitFromViper initializes Options with properties from viper
 func (opt *Options) InitFromViper(v *viper.Viper) {
 	opt.config = producer.Configuration{
-		Brokers: strings.Split(stripWhiteSpace(v.GetString(configPrefix+suffixBrokers)), ","),
+		Brokers:          strings.Split(stripWhiteSpace(v.GetString(configPrefix+suffixBrokers)), ","),
+		RequiredAcks:     v.GetInt(configPrefix + suffixRequiredAcks),
+		CompressionLevel: v.GetInt(configPrefix + suffixCompressionLevel),
```
Comment (objectiser) [inline plugin/storage/kafka/options.go]: You are not setting `Compression`.

File: plugin/storage/kafka/options_test.go [hunk 4]
```diff
@@ -28,12 +28,18 @@ func TestOptionsWithFlags(t *testing.T) {
 	command.ParseFlags([]string{
 		"--kafka.producer.topic=topic1",
 		"--kafka.producer.brokers=127.0.0.1:9092, 0.0.0:1234",
-		"--kafka.producer.encoding=protobuf"})
+		"--kafka.producer.encoding=protobuf",
+		"--kafka.producer.required.acks=1",
+		"--kafka.producer.compression=0",
+		"--kafka.producer.compression.level=-1000"})
```
Comment (objectiser) [inline plugin/storage/kafka/options_test.go]: Should use non-default values for these properties, to ensure they are being set.

File: plugin/storage/kafka/options_test.go [hunk 5]
```diff
@@ -28,12 +28,18 @@ func TestOptionsWithFlags(t *testing.T) {
 	command.ParseFlags([]string{
 		"--kafka.producer.topic=topic1",
 		"--kafka.producer.brokers=127.0.0.1:9092, 0.0.0:1234",
-		"--kafka.producer.encoding=protobuf"})
+		"--kafka.producer.encoding=protobuf",
+		"--kafka.producer.required.acks=1",
```
Comment (objectiser) [inline plugin/storage/kafka/options_test.go]: This is still using the default value - suggest setting to 0.

File: plugin/storage/kafka/options.go [hunk 6]
```diff
@@ -69,12 +75,30 @@ func (opt *Options) AddFlags(flagSet *flag.FlagSet) {
...
+	)
+	flagSet.Int(
+		configPrefix+suffixCompression,
+		defaultComression,
+		"(experimental) Type of compression to use on messages. default = 0, none = 0, gzip = 1, snappy = 2, lz4 = 3, zstd = 4",
+	)
+	flagSet.Int(
+		configPrefix+suffixCompressionLevel,
+		defaultCompressionLevel,
+		"(experimental) Level of compression to use on messages. default= -1000, gzip = 1-9 (default = 6), snappy = none, lz4 = 1-17 (default = 9), zstd = -131072 - 22 (default = 3)",
```
Comment (objectiser) [inline plugin/storage/kafka/options.go]: Sorry didn't pick up on this before, but there isn't a single default for this property - there is a default per compression type. So the default will need to be determines based on the value of the `compression` flag. And then have a test that checks the default for each of the valid `compression` values.

Comment (chandresh-pancholi) [reply plugin/storage/kafka/options.go]: I have followed below kafka documentation for compression level.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-390%3A+Allow+fine-grained+configuration+for+compression
Default values are based on current compression libraries followed in Kafka.

File: plugin/storage/kafka/options_test.go (hunk 5 above)
Comment (chandresh-pancholi) [reply plugin/storage/kafka/options_test.go]: Apologies. I will make the change.

**Pull request closed without merging**